### PR TITLE
Grafana-UI: Refactor legacy inline form components

### LINKED ISSUE #27596: Grafana-UI: Refactor legacy inline form components
As discussed in the design group meeting it'd be a good idea to start refactoring the legacy inline form components, even though no redesigns have been done yet. As long as the api is kept consistent the future design changes shouldn't affect the components usage.

**What would you like to be added**:

- [ ] Refactor `.gf-form` components into `InlineField` (or `Field` with `inline` toggle prop), similar to the current `Field` component.
- [ ] Move all the related styling from `.scss` to emotion.
- [ ] Make sure the components look and behave properly inside `Form`.

**Why is this needed**:
- Ensure consistent usage of components
- Simplify html structure by providing wrapper components 
- Improve accessibility by adding some attribute automatically (for ex. adding `htmFor` to the label based on input's `id`)

### PR DESCRIPTION
**Pull request by Clarity-89**
**What this PR does / why we need it**:
- Add `InlineField` component (refactor of legacy `FormField`)
- Add `InlineLabel` component (basically moved the current `InlineFieldLabel` to the `Forms`, so not to break legacy compatibility)
- Add `InlineFieldRow` to align inline fields in rows (works save way as current `gf-form-inline`).

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/27596

Screenshots:
Screenshot 2020-09-21 at 15 33 17

### CODE REVIEW

**Comment by peterholmberg** [comment]
The tooltip is a bit off :thinking: 
Screenshot from 2020-09-22 10-28-43

**Comment by Clarity-89** [comment]
> The tooltip is a bit off 🤔
> Screenshot from 2020-09-22 10-28-43

Yeah, it seems like it's off for all the components in storybook, couldn't figure out what's wrong.

File: packages/grafana-ui/src/components/Forms/InlineField.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  className?: string;
+  /** Make the field's child to fill the width of the row. Equivalent to setting `flex-grow:1` on the field */
+  grow?: boolean;
+  /** A toggle to apply query keyword styling to the label */
+  isKeyword?: boolean;
+  /** Fill the remaining width of the row with the label's background */
+  fill?: boolean;
+}
+
+export const InlineField: FC<Props> = ({
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/InlineField.tsx:33]: Shouldn't this extend [FieldProps](https://github.com/grafana/grafana/blob/416111c5f6afb32c4036724d7f87fa7a1102055e/packages/grafana-ui/src/components/Forms/Field.tsx#L8)? I think the HTMLProps extension should belong to the FieldProps too, then you could use that type as a base for InlineField properties

File: packages/grafana-ui/src/components/Forms/InlineFieldRow.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import React, { FC, ReactNode, HTMLProps } from 'react';
+import { css, cx } from 'emotion';
+import { useStyles } from '../../themes';
+
+export interface Props extends Omit<HTMLProps<HTMLDivElement>, 'css'> {
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/InlineFieldRow.tsx:5]: Hmm, why `css` is omitted here? How come is this a property on HTMLProps type?

File: packages/grafana-ui/src/components/Forms/InlineLabel.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import React, { FunctionComponent, HTMLProps, ReactNode } from 'react';
+import { GrafanaTheme } from '@grafana/data';
+import { css, cx } from 'emotion';
+import { Tooltip, PopoverContent } from '../Tooltip/Tooltip';
+import { Icon } from '../Icon/Icon';
+import { useTheme } from '../../themes';
+
+export interface Props extends Omit<HTMLProps<HTMLLabelElement>, 'children' | 'className' | 'css'> {
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/InlineLabel.tsx:8]: Think this should extend [Label props](https://github.com/grafana/grafana/blob/416111c5f6afb32c4036724d7f87fa7a1102055e/packages/grafana-ui/src/components/Forms/Label.tsx#L8) to keep the interfaces in sync. WDYT?

**Review by dprokop — COMMENTED** [review]
Looks good, left few nits to consider :)

File: packages/grafana-ui/src/components/index.ts [hunk 4]
```diff
@@ -145,6 +145,8 @@ export { Field } from './Forms/Field';
 export { Legend } from './Forms/Legend';
 export { FieldSet } from './Forms/FieldSet';
 export { FieldValidationMessage } from './Forms/FieldValidationMessage';
+export { InlineField } from './Forms/InlineField';
```
Comment (Elfo404) [inline packages/grafana-ui/src/components/index.ts:148]: Shouldn't `InlineFieldRow` also be exported here?

```suggestion
export { InlineField } from './Forms/InlineField';
export { InlineFieldRow } from './Forms/InlineFieldRow';
```

Comment (Clarity-89) [reply packages/grafana-ui/src/components/index.ts:148]: Good catch! Missed that one.

File: packages/grafana-ui/src/components/Forms/InlineFieldRow.tsx (hunk 2 above)
Comment (Clarity-89) [reply packages/grafana-ui/src/components/Forms/InlineFieldRow.tsx:5]: For some reason it ends up in the component's Prop table, which is confusing for the users. Could be related to https://github.com/grafana/grafana/issues/26512 maybe?

File: packages/grafana-ui/src/components/Forms/InlineLabel.tsx (hunk 3 above)
Comment (Clarity-89) [reply packages/grafana-ui/src/components/Forms/InlineLabel.tsx:8]: Good point! Haven' thought about that.

File: packages/grafana-ui/src/components/Forms/InlineField.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  /** Content for the label's tooltip */
+  tooltip?: PopoverContent;
+  /** Custom width for the label */
+  labelWidth?: number | 'auto';
+  /** Make the field's child to fill the width of the row. Equivalent to setting `flex-grow:1` on the field */
+  grow?: boolean;
+  /** A toggle to apply query keyword styling to the label */
+  isKeyword?: boolean;
+  /** Fill the remaining width of the row with the label's background */
+  fill?: boolean;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Forms/InlineField.tsx:19]: what's the difference between grow and fill? Feels like we only need one of them?

Comment (Clarity-89) [reply packages/grafana-ui/src/components/Forms/InlineField.tsx:19]: `grow` fills the rest of the row with a form element (input, etc):
Screenshot 2020-09-24 at 14 07 50
`fill` fills the rest of the row with dark background:
Screenshot 2020-09-24 at 14 07 57
But I also heard that we might be getting rid of the fill background, so I'm waiting for @jessover9000 feedback on this PR.

File: packages/grafana-ui/src/components/Forms/InlineField.mdx [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { Props } from "@storybook/addon-docs/blocks";
+import { InlineField } from "./InlineField";
+
+# InlineField
+
+A basic component for rendering inline form elements together with their label. If the child element has `id` specified, the label's `htmlFor` attribute, pointing to the id, will be added.
```
Comment (jessover9000) [inline packages/grafana-ui/src/components/Forms/InlineField.mdx:6]: Nice starting point. I think it's useful to list which inline form elements you could use to render with the label, and also give related component names here for 100% clarity (e.g. not just say "together with their label" but "together with `InlineLabel`", mention `InlineFieldRow`).

> If the child element has `id` specified, the label's `htmlFor` attribute, pointing to the id, will be added.
The phrasing is a bit complicated; maybe say "…specified, the component automatically gets a `htmlFor` attribute which points to the id". Also I don't yet understand why this information is important or why the label needs an attribute pointing to the id.

Maybe it's a good idea to explain how you can tweak the component to fit your needs, e.g. that you can modify the width of labels and/or fields using [list of all props that do this], what it looks like when the children prop is used, etc.

Can I have one label with multiple fields? If yes, how do I do that?

File: packages/grafana-ui/src/components/Forms/InlineFieldRow.mdx [hunk 7]
```diff
@@ -0,0 +1,14 @@
+Used to align multiple inline form fields in one row. The row will wrap if the width of the children exceeds its own. Equivalent to the div with `gf-form-inline` class name.
```
Comment (jessover9000) [inline packages/grafana-ui/src/components/Forms/InlineFieldRow.mdx:1]: > The row will wrap if the width of the children exceeds its own.
I kind of know what you mean but this is still a bit confusing to me. Maybe change to "If the width of the children exceeds the row width, the row will wrap." Also I guess a screenshot example of how wrapped rows look would be nice. 

You explain that a row stacks multiple fields, but you don't address using multiple rows and it's not in the code example either. I guess it's fine to just put multiple rows one after another? Also there is no props table for this component, is that on purpose?

Just like in the InlineField docs, I would list related components and how they can and should be used within this component.

File: packages/grafana-ui/src/components/Forms/InlineLabel.mdx [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { Props } from "@storybook/addon-docs/blocks";
+import { InlineLabel } from "./InlineLabel";
+
+# InlineLabel
+
+A horizontal variant of Label, primarily used in query editors.
```
Comment (jessover9000) [inline packages/grafana-ui/src/components/Forms/InlineLabel.mdx:6]: I feel like the limitations of the InlineLabel could be explained a bit more here: 
- There is no description option, if you need to add additional explanation use the tooltip variant. 
- For query editor readability, the label text should be as short as possible (4 words or less).
- Which components can the label be combined with and how does that work (`InlineField`)?

**Review by jessover9000 — COMMENTED** [review]
I added some feedback for the docs text. I think this is a great start!

Regarding the fill and grow options, as mentioned in the last UX feedback session, I recommend getting rid of both. These options make add and remove buttons move to the very right of the query editor for many data sources, see this Elastic example:
image
On wide screens, the fill/grow principle would make the distance between the majority of query editor components and these buttons very big, and users are likely to miss elements that are too far away from their center of perceptions. The grow principle makes sense for inputs that we expect very long code input for, but should be used sparingly. (that's something to add to the docs ;) )

If we get rid of the fill principle, the only full-width elements would be the top rows separating queries. The whitespace could make it much easier to scan where your query ends and the next one starts. See screenshots for comparison:
image
image

Regarding font colors of the components, I think the white label text works better than the previous blue. I would make the labels Regular weight instead of Medium, to make the clickable parts stand out more. You are using form elements instead of the button-like field design in live Grafana, so the color tests I've done don't match 100% right now, but here's the team favourite from yesterday's UX feedback session: https://www.figma.com/file/ITCsaXtsANM17oRtU6GQS9/Component-Playground?node-id=672%3A6

File: packages/grafana-ui/src/components/Forms/InlineFieldRow.mdx (hunk 7 above)
Comment (Clarity-89) [reply packages/grafana-ui/src/components/Forms/InlineFieldRow.mdx:1]: Yeah, there's no props for this component except the native HTML ones and storybook doesn't display those for some reason.

**Review by jessover9000 — APPROVED** [review]
Added a small nit, otherwise looks good! 💪

**Pull request merged**